### ISSUE TITLE
Helm 'delete' doesn't delete PVCs 

### ISSUE THREAD

**Issue by juan131** [helm/helm#5156]
When Helm installs a chart including a **statefulset** which uses **volumeClaimTemplates** to generate new PVCs for each replica created, Helm loses control on those PVCs. 

As a consequence, once the chart is removed (both using `helm delete my-release` and `helm delete --purge my-release`) every PVC created is left on the cluster.

Helm should take care of those resources to maintain the cluster "clean".

Output of `helm version`: **2.12.1** (latest)

Output of `kubectl version`: **1.12.4**

Cloud Provider/Platform (AKS, GKE, Minikube etc.): All of them

**Comment by desaintmartin** [comment]
While I don't think it should be deleted by default by a `helm delete --purge`, there should be some flag to clean up generated kube objects like those PVCs.

**Comment by bacongobbler** [comment]
Agreed. PVCs is one of those resources that don't get cleaned up by `helm delete --purge`, and I believe that's been the intended behaviour since v2.0.0 in order to protect that data in case an operator needed to purge and re-install. A feature flag to delete *all* of the resources managed by a helm release could be useful in certain cases.

**Comment by juan131** [comment]
Yes, I agree
Maybe `--purge` it's not the proper tag but an alternative one should be implemented instead.

**Comment by koulq** [comment]
Is there any good plan for this question now? Why isn't anyone answering?

**Comment by RobertoDonPedro** [comment]
Personally I think that the helm behaviour should be in line with the corresponding kubernetes behaviour in this regard.
If you define PVCs and PVs by `kubectl apply ...` and then run `kubectl delete ...` the previousely defined PVCs and PVs get deleted again. 
In comparison to that, if you run `helm install ...` and then `helm del --purge ... ` the previousely defined PVCs and PVs stay intact. 
Although these commands are not comparable one to one, I would opt for also having an intuitive behaviour available with helm.

**Comment by ALuongIBM** [comment]
Interesting contrast @RobertoDonPedro, perhaps we should remind the user that PVC's exist or even offer an option like `--purge-pv=true`?

**Comment by bacongobbler** [comment]
> Is there any good plan for this question now? Why isn't anyone answering?

What do you propose as a solution? :)

At this point it appears there's consensus here that Helm should follow the same behaviour as `kubectl` here. If someone wants to write a PR against Helm 3 to remove the PVC on `helm delete`, we'd appreciate it. Users that rely on the PVC staying after a `helm delete` can use the `helm.sh/resource-policy: keep` annotation.

**Comment by winjer** [comment]
i have helm releases where i want to ensure the pvcs do not get deleted in case a release is deleted, since they contain critical data. it would be really good for the behaviour to be clear at least - if deployment pvcs do get deleted but statefulset pvcs do not then that would be good to understand.

**Comment by pxLi** [comment]
would be nice to just honor PV's `reclaim policy` :)

**Comment by bacongobbler** [comment]
At this point we cannot change the behaviour here until Helm 4. Tagging it as a proposal we can look at for Helm 4.

**Comment by hetii** [comment]
> would be nice to just honor PV's `reclaim policy` :)

Hi.
This is my  workaround that I use in helmfile:

     # Remove pvc with reclaim policy == delete.
      - events: ["postuninstall"]
        command: "bash"
        showlogs: true
        args: ["-c", "kubectl -n {{`{{.Release.Namespace}}`}} get pvc -l app.kubernetes.io/instance={{`{{.Release.Name}}`}} -o jsonpath='{.items..spec.volumeName}' | xargs -n1 -d ' ' -I{} -r kubectl get pv {} -o json | jq -r 'select(.spec.persistentVolumeReclaimPolicy==\"Delete\")|.spec.claimRef.name' | xargs -I{} -r kubectl -n {{`{{.Release.Namespace}}`}} delete pvc {}"]

You can also use below example and wrap it as a bash function:

    kubectl -n mariadb get pvc -l app.kubernetes.io/instance=common-galera -o jsonpath='{.items..spec.volumeName}' | KUBECONFIG=/home/dev/.kube/path-to-your-config xargs -n1 -d ' ' -I{} -r kubectl get pv {} -o json | jq -r 'select (.spec.persistentVolumeReclaimPolicy=="Delete")|.spec.claimRef.name'| KUBECONFIG=/home/dev/.kube/path-to-your-config xargs -I{} -r kubectl -n mariadb delete pvc {}

The KUBECONFIG is optional.

**Comment by slmjy** [comment]
Just a funny consideration - I accidentally ran helm uninstall today on a wrong kubernetes context and this issue being NOT fixed SAVED MY LIFE :) I was able to run helm install, and all the stateful services were back, no data loss!   So, please, until uninstall behavior remains as it is now - do not fix this!

**Comment by bacongobbler** [comment]
Kinda the reason why we implemented it that way. ^

**Comment by jastabile** [comment]
@bacongobbler and its the best decision. 

But can you add a flag as other people said?

**Comment by bacongobbler** [comment]
Please feel free to work on something - I have way too much on my plate at the moment to look at this. But we'd welcome contributions.

**Comment by jetndra** [comment]
hi, 
  Does anyone try with resource-policy annotation for sts pvc delete ?
  
kind: StatefulSet
metadata:
  annotations:
   **helm.sh/resource-policy": delete-on-completion**

**Comment by melissachang** [comment]
I was not so lucky as @slmjy . Based on this issue, I ran `helm uninstall`, thinking my PVCs would persist. Instead, some of my PVCs were deleted. :( I filed #10820. Maybe I'm misunderstanding something.

**Comment by Eternal21** [comment]
> 

I am trying to use your workaround, but do not see `postuninstall` hook in helm?  Will `post-delete` work for this?  https://helm.sh/docs/topics/charts_hooks/

**Comment by ghostdevhv** [comment]
@bacongobbler - Not sure if you are working on this project, but do we have any solution for this? 

I am installing the Prometheus and attaching the PV with a few pods and I want `helm delete` to delete the PVs along with the chart. Currently, the volume is not getting deleted

**Comment by bhack** [comment]
This is going to impact also the new Bitnami MLflow chart.
/cc @aoterolorenzo

**Comment by cornfeedhobo** [comment]
## TL;DR:
- The current behavior is expected for safety. If you delete a chart, _or do a force upgrade_, your data doesn't disappear.
- Staff are too busy to prioritize this; most production environments don't want this behavior.
- This behavior would be useful for specific development use cases
- No one has started a PR
- If someone starts a PR, it should be a new flag that enables deleting.
- This will not be backported to 3.x, but if no one makes a PR soon, it also won't make it into 4.x

**Comment by bhack** [comment]
@cornfeedhobo I am ok if it is the expected behavior but do we have a documentation for the best practices to notify this behavior to the users in the the chart logic? In many charts this error is very obscure to identify.

**Comment by cornfeedhobo** [comment]
@bhack It would be on the chart creator to write something specific to what they've made. If they know the chart will make PVs, they can include some [notes that get outputted at install time](https://helm.sh/docs/chart_template_guide/notes_files/), or they can explore using a chart value and [helm hooks](https://helm.sh/docs/topics/charts_hooks/). I would also suggest anyone search github for creative examples of these - I've seen some interesting hook scripts over the last few years.

**Comment by alexey-cord-tech** [comment]
@cornfeedhobo I don't see this mentioned in this thread, but I would argue that this *is* a bug when you consider how this interacts with `--wait` (which is, "if a chart has PVCs then `delete --wait` will always time out with an error").

This breaks the obvious workflow of using Helm to run one-off jobs (migrations, performance tests, this kind of stuff): jobs are immutable so the "run again" has to be "uninstall then install again". And then if the chart is anything more complicated than a single POD, then it fails with "stuff is still being deleted, can't install". And there's no generic way to add a "only run install once uninstall is completed" because of the above: `--wait` will fail if there's a PVC involved.

Looks like one of the two has to give: either `delete --wait` should not be waiting for the stuff that's left behind on purpose (at least optionally, `--ignore-storage` or smth?); or there should be a `--force-delete-storage` kind of a flag.

Currently people resort to manual-ish shenanigans with "kubectl delete based of labels" which are *much less safe than a built-in behaviour would be*.
